### PR TITLE
Regex support in publisher ACL rules.

### PR DESCRIPTION
**Pull request by DmitryKuzmenko**
### What does this PR do?

Users in publisher ACL and users and functions in the publisher ACL
blacklist now have matched in the following order:
1. Equality
2. Glob
3. Regex

Users in publisher ACL still not supported in systems without pwd python
module (Windows, any else?)
### What issues does this PR fix or reference?
saltstack/salt#23194
### Tests written?

Partially.

Please don't merge for now, I've to test it and write some tests.

### CODE REVIEW

**Comment by DmitryKuzmenko** [comment]
@cachedout could you please take a look just to confirm I'm in the right way.
Probably it's better to not use the `salt.utils.expr_match` here and just use more strict `re.match`?

File: salt/daemons/masterapi.py [hunk 1]
```diff
@@ -187,11 +187,51 @@ def clean_old_jobs(opts):
...
+        if salt.utils.is_windows() and not os.access(keyfile, os.W_OK):
+            # Cannot delete read-only files on Windows.
+            os.chmod(keyfile, stat.S_IRUSR | stat.S_IWUSR)
+        os.unlink(keyfile)
+
+    key = salt.crypt.Crypticle.generate_key_string()
+    cumask = os.umask(191)
+    with salt.utils.fopen(keyfile, 'w+') as fp_:
+        fp_.write(key)
+    os.umask(cumask)
```
Comment (cachedout) [inline salt/daemons/masterapi.py]: Do we want to restore the umask after we set it? That's typically what we do.

File: salt/utils/__init__.py [hunk 2]
```diff
@@ -1333,47 +1333,39 @@ def check_whitelist_blacklist(value, whitelist=None, blacklist=None):
...
-    in_blacklist = False
+    if blacklist is not None:
+        if not hasattr(blacklist, '__iter__'):
+            blacklist = [blacklist]
+        try:
+            for expr in blacklist:
+                if expr_match(value, expr):
+                    return False
+        except TypeError:
+            log.error('Non-iterable blacklist {0}'.format(whitelist))
```
Comment (cachedout) [inline salt/utils/__init__.py]: Did you mean for the `format()` call to contain `whitelist` here?

**Comment by cachedout** [comment]
So long as we maintain the ordering that you describe between equality, glob and regex I think it is fine to use either `re` or `expr_match`. What do you think, @thatch45 and @terminalmage ?

**Comment by thatch45** [comment]
I think that it using expr_match is the right way to do it because it maintains consistent behavior with other similar interfaces in salt

File: salt/daemons/masterapi.py (hunk 1 above)
Comment (DmitryKuzmenko) [inline salt/daemons/masterapi.py]: @cachedout here we change umask to create the key file read-only for the owner only. Then we restore our umask to return the normal behavior for other files. This looks correct for me.
Also this code wasn't touched by me, I've just moved it into a separate function.

File: salt/utils/__init__.py (hunk 2 above)
Comment (DmitryKuzmenko) [inline salt/utils/__init__.py]: Sure! Fixed, thank you!

**Comment by terminalmage** [comment]
Yeah, expr_match is used in a bunch of other places and is an established way of doing matching, I like the use of expr_match as well.

**Comment by cachedout** [comment]
@DmitryKuzmenko Let me know when this is ready to go from your end, please. Thanks!

**Comment by DmitryKuzmenko** [comment]
Go-go Jenkins!

**Comment by DmitryKuzmenko** [comment]
@cachedout it's done. Just let's wait for tests pass.

**Comment by cachedout** [comment]
@DmitryKuzmenko Wonderful. Thank you!

**Comment by thatch45** [comment]
@cachedout I think this looks good to go

**Comment by cachedout** [comment]
I think this broke the LocalClient:

```
>>> import salt.client
>>> o = salt.config.DEFAULT_MASTER_OPTS
>>> o['id'] = 'testing'
>>> c  = salt.client.get_local_client(mopts=o)
>>> c.cmd('*', 'saltutil.running', timeout='5')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    c.cmd('*', 'saltutil.running', timeout='5')
  File "/home/mp/devel/salt/salt/client/__init__.py", line 560, in cmd
    **kwargs)
  File "/home/mp/devel/salt/salt/client/__init__.py", line 315, in run_job
    raise SaltClientError(general_exception)
SaltClientError: 'str' object has no attribute 'pop'
>>> c.cmd('*', 'saltutil.running', timeout='5')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    c.cmd('*', 'saltutil.running', timeout='5')
  File "/home/mp/devel/salt/salt/client/__init__.py", line 560, in cmd
    **kwargs)
  File "/home/mp/devel/salt/salt/client/__init__.py", line 317, in run_job
    return self._check_pub_data(pub_data)
  File "/home/mp/devel/salt/salt/client/__init__.py", line 242, in _check_pub_data
    'Failed to authenticate! This is most likely because this '
EauthAuthenticationError: Failed to authenticate! This is most likely because this user is not permitted to execute commands, but there is a small possibility that a disk error oc
curred (check disk/inode usage).
```

**Comment by DmitryKuzmenko** [comment]
@cachedout I've compared the code related to sudo users before this patch and found that I've really broke the logic. Please review my #34808 PR: it has 100% the same logic as before but with regex support.

**Pull request merged**